### DISCUSSION TITLE
Sharepoint Library - xlsx files not downloading at first attempt

### DISCUSSION THREAD

**Discussion by tdcockers** [abraunegg/onedrive#1667] [General]
I'm trying to sync a Sharepoint document library, but the first time I try to sync using the --monitor switch any xlsx files are outputting "ERROR: File download size mis-match. Increase logging verbosity to determine why." Switching to debug output doesn't appear to provide any additional useful information.

Doing a --resync afterwards appears to have no issues with the files that have been scanned previously, but new files still give the same errors.

Environment is Nethserver 7.9 built per instructions for Centos 7 on Github, Everything seems to be working fine other than this.

Any suggestions?
Thanks!

**Comment by abraunegg** [comment]
@tdcockers 
> "ERROR: File download size mis-match. Increase logging verbosity to determine why." Switching to debug output doesn't appear to provide any additional useful information.

This is because Microsoft modifies all PDF, MS Office & HTML files with added XML content. It is a 'feature' of SharePoint. This effectively changes the size of the file when files are on SharePoint. There is no way to turn this off or disable the way Microsoft wrote SharePoint. Review https://github.com/OneDrive/onedrive-api-docs/issues/935  for further details.

As to your issue:

What local filesystem is being used on Nethserver 7.9 where the data is being saved? It could also be that a local process is 'scanning' or mangling the downloaded file .. given that your platform has integrated AntiVirus protection via ClamAV - so essentially a race condition between the file being downloaded & scanned before it is fully written to disk (hence the error).

> Doing a --resync afterwards appears to have no issues with the files that have been scanned previously, but new files still give the same errors.

Dont use --resync unless the application tells you to ... this deletes your 'local known state' and just re-validates your data, however if there is no issue with existing files that previously threw that error, then the AntiVirus / ClamAV race is a plausible item - because if the file truly was different on disk, you still would get the file size error message.

With your specific error it would be actually interesting to review a full debug log for this issue to give you a full complete picture of what is occurring. This will confirm that the full file was downloaded, and it is the disk reporting post file being written to disk where the issue potentially is.

As a test - disable ClamAV and re-run your tests - it would be interesting to see if there is any difference here.

**Comment by tdcockers** [comment]
OK, makes sense thanks. When the error was being thrown, there was nothing left on the disk; only files that didn't error out were written to disk. I don't have any antivirus or the like running on this system at the moment, but started thinking about the filesystem... NS recommends XFS,

As a test I re-formatted it as ext4. Interestingly, when I re-run the download, everything is fine up to the point it got to last time (2016 documents in this case), but then when it starts on 2017 documents it starts presenting the error again. Without knowing how it works in the background, it sounds like the file is being checked before the database entry is recorded or something quirky like that?

Let me know what I can do or send if you want to look into it further.

**Comment by abraunegg** [comment]
@tdcockers 
> As a test I re-formatted it as ext4, and the first run seems to not be having any errors. maybe the journalling of XFS was causing the race condition?
> 
> As a test I re-formatted it as ext4. Interestingly, when I re-run the download, everything is fine up to the point it got to last time (2016 documents in this case), but then when it starts on 2017 documents it starts presenting the error again.

Possible - however my *primary* development system for this client is CentOS (CentOS Linux release 7.9.2009 (Core)) - with all filesystems running XFS - and I certainly do not run into this sort of issue, despite rather extensive testing all sorts of user account types + SharePoint. If this was an issue with XFS, I most likely would have seen this | found this a long time ago.

What kernel version is your 'Nethserver 7.9' running ? Is this the stock standard CentOS kernel given they are using this as a base - or have they compiled | built their own?

I would potentially start looking deeper as well - maybe your hardware | platform | disks | disk controller | cables ... yes everything *might* be working - but you also *might* have some sort of underlying hardware issue that is only highlighted when using your chosen platform ... 

To rule this out, can you install a pure | plain CentOS OS on the exact same hardware and see what happens when it is just CentOS + XFS ? If you do not have an issue in that scenario - then, that leaves *something* in your choice of platform as the cause of your issue - hopefully Nethserver has platform support for which you can leverage this this is the case. If the issue is still present when using a pure CentOS install - then, I would look harder at your hardware.

Some other areas to look at:
* Check all your physical media for reallocated sectors looking at the RAW_VALUE item
  * Example: `smartctl -a /dev/sda | grep Reallocated_Sector_Ct`
* Check all your physical media for pending sectors
  * Example: `smartctl -a /dev/sda | grep Current_Pending_Sector`

The only acceptable value here for RAW_VALUE for my disks is '0' .. you may take a view that any value between 1 and 10 is acceptable. Either way, that points to physical media issues = disk issues = disk write problems. If you have disks with issues, I personally would replace them.

> Without knowing how it works in the background, it sounds like the file is being checked before the database entry is recorded or something quirky like that?

The way the client works is that the file is downloaded & written to disk. Once the file write is complete, the new file that was written is queried to get the size of the file on disk, and compared with that of the file JSON data from OneDrive. If there is a size mismatch, this error is presented to inform you that what you downloaded - is not equal on your disk.

That both XFS + EXT4 is reporting an incorrect filesystem size is concerning.

*Something* on your platform - hardware or your choice of system is contributing to this issue you are facing.

*If* you are still evaluating solutions, look at systems that use ZFS for your data file systems rather that ext4, xfs and so on - much better data integrity & reliability .. and handles disk writes of data very differently. If your hardware is the contributing cause - maybe ZFS will provide you a way to get around your hardware issues for the moment.

**Comment by tdcockers** [comment]
Thanks! Will dig/test further and get back to you. I'm running NS as a VM on XCP-ng rather than bare metal.

Kernel is 3.10.0-1160.45.1.el7.x86_64, I think it is a straight downstream of Centos 7.9.2009 but they do use some custom repos that may have different versions of some utils.

**Comment by abraunegg** [comment]
> Thanks! Will dig/test further and get back to you. I'm running NS as a VM on XCP-ng rather than bare metal.

In this case - your *issue* is also potentially related to XCP-ng and how it is presenting the disks to your Virtual Machine .... now that you have detailed that ... I am almost *certain* this will be your issue ...

**Comment by tdcockers** [comment]
OK, getting back to this one with more detail. It's possibly an xcp/environment thing as I haven't tried outside of that yet, but I'm not sure why.

First clarification is that it seems to be any office files, not just excel. PDF and other app files are fine, but anything microsoft (xls, xlsx, doc, docx etc) is not working.

As a test, I did a clean install of Debian in the same VM as Nethserver (full wipe and reinstall), wiped the drive that was holding the synced data, and synced onedrive again. Everything went fine and seemed to be working without errors... but it seems like once it got to the files that hadn't been synced previously (in the old VM), the errors started again.

I'm thinking there must be something in Sharepoint causing issues, perhaps related the the metadata that Microsoft adds per your earlier link. If the files that are touched the first time fail but work the second time, even if the second time is from essentially a different machine, I'm thinking the only common reference between them is Sharepoint.

These are mostly quite old files, as a hypothesis is it possible that Sharepoint is marking unused Microsoft documents as 'archived' or something like that, but when they are touched it removes the flag and creates the discrepancy?

Separately but possibly related, when the sync started and the folder tree was created, it seems like a lot of the newer folders weren't created. I'm not at all sure (or competent to figure out) how the sync works, does it sync files and folders in the order they were created (which makes sense to me because I migrated a whole lot of stuff to Sharepoint at the same time, before using it as a daily driver)?

I haven't let the sync run all the way through yet to test, let me know if you still prefer the environment as the likely problem and I'll try it on a different system altogether; otherwise I can let it run and see if it comes good later in the process.

**Comment by abraunegg** [comment]
@tdcockers 

> I'm thinking there must be something in Sharepoint causing issues, perhaps related the the metadata that Microsoft adds per your earlier link. If the files that are touched the first time fail but work the second time, even if the second time is from essentially a different machine, I'm thinking the only common reference between them is Sharepoint.
>
> These are mostly quite old files, as a hypothesis is it possible that Sharepoint is marking unused Microsoft documents as 'archived' or something like that, but when they are touched it removes the flag and creates the discrepancy?

This is quite possible ... SharePoint does some very odd stuff - but I have not seen in ~4 years anything like you are describing.

> Separately but possibly related, when the sync started and the folder tree was created, it seems like a lot of the newer folders weren't created. I'm not at all sure (or competent to figure out) how the sync works, does it sync files and folders in the order they were created (which makes sense to me because I migrated a whole lot of stuff to Sharepoint at the same time, before using it as a daily driver)?
> 
> I haven't let the sync run all the way through yet to test, let me know if you still prefer the environment as the likely problem and I'll try it on a different system altogether; otherwise I can let it run and see if it comes good later in the process.

I have not seen any behavior like what you are describing. The client, when operating, will query the OneDrive API and pull down all the changes as per the API, based on the information the API is presenting. In your case, it sounds like the /Delta link is not operating correctly - or, due to the disk issues you are seeing - something is really messing up how this is being handled.

I strongly suspect that a lot of this is related to how your disks & underlying disk storage is being presented to your virtual machine.

Is your XCP-ng system installed on *your* hardware that you have access to - or are you running XCP-ng on a cloud provided system?

I would start by removing any complexity at all. If you have a simple laptop | desktop running Linux that is not being virtualised in any way, run the client there so that you can see what is going on - giving you a true reference point.

If you are running XCP-ng on your own hardware - server at home for example - I strongly suggest that you switch to use VMware ESXi - it is free after all (it is vCentre / management where all the costs come in). I use VMware ESXi here for all my stuff. The local storage is presented as VMFS volumes where I can create whatever virtual machines using local storage. For other testing I have a NAS where I can present iSCSI or NFS volumes .. any virtual machine I create - from Windows to Linux to BSD to Solaris to even systems running QEMU to emulate RaspberryPi or Mainframe (using Hurcules) - never once have I encountered an issue like what you are seeing.

If you are running XCP-ng on a cloud provider platform ... then you really need to go back and understand how they are presenting disks, how XCP-ng interprets this (block write holes, thin provision, allocations etc) and how then XCP-ng is creating your VM file systems & how then the file system in the VM is reacting to the different storage layers.

**Comment by abraunegg** [comment]
@tdcockers 
Any update here re testing your environment?

**Reply by tdcockers** [reply]
Sorry haven't had time to set it up yet... will keep you posted.

**Reply by tdcockers** [reply]
OK, update at last... bare metal Debian 11 install, OD compiled from source... same behaviour as [previous reply](https://github.com/abraunegg/onedrive/discussions/1667#discussioncomment-1556646).

At this point I can't see that it is anything other than a Sharepoint quirk... anything you want me to try?

**Reply by abraunegg** [reply]
@tdcockers 
A full debug log sent via email would be beneficial to see if there is something less obvious going on

**Reply by tdcockers** [reply]
Is that using --verbose --verbose?

I can't seem to get it to produce a log either, I've set the log directory but nothing is being sent there.

**Reply by abraunegg** [reply]
@tdcockers 
It all comes down to 'how' you are running the application.

To generate a debug log, yes it is `--verbose --verbose` - but if you are running it via systemd or some other process then it is going to be slightly different.

To generate the debug log - follow this process: https://github.com/abraunegg/onedrive/wiki/Generate-debug-log-for-support

**Reply by tdcockers** [reply]
Thanks, sent.

**Comment by abraunegg** [comment]
@tdcockers 
Thanks for all the debug logs.

Please can you test the following PR (#1686) to assist with the issue you are facing.

```
git clone https://github.com/abraunegg/onedrive.git
cd onedrive
git fetch origin pull/1686/head:pr1686
git checkout pr1686
./configure; make clean; make;
```
To run the PR, you need to run the client from the PR build directory:
```
./onedrive <any options needed> --disable-download-validation
```

When running the PR, your version should be `onedrive v2.4.13-25-ge0f999b`

**Reply by tdcockers** [reply]
Thanks, that seems to be working well, except for some timestamp conflicts between Sharepoint and files that I've previously downloaded during testing to this machine. I'll spin up a new one and try a clean setup on that and let you know how it goes.

I'm thinking that perhaps the disable validation switch should be a first-run thing only, and once Sharepoint has pulled all the files out of cold storage or wherever it is holding them that the switch should be unnecessary when running in future after that?

**Comment by abraunegg** [comment]
@tdcockers 
> I'm thinking that perhaps the disable validation switch should be a first-run thing only, and once SharePoint has pulled all the files out of cold storage or wherever it is holding them that the switch should be unnecessary when running in future after that?

Correct. I will update the PR, so that *if* this gets triggered, the output message will say to use the disable validation flag.

Outside of this, there is not much that can be done except flag an issue with the OneDrive API as this is a OneDrive API | SharePoint bug
